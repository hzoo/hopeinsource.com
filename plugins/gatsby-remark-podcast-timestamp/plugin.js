const visit = require("unist-util-visit");
const toString = require("mdast-util-to-string");
const timestampRegex = /(\d+:\d+(:\d+)?)/;

function isTimestamp(p) {
  // has to be the first child in a paragraph and match the regex
  // [00:28] **Speaker**: more text here
  return (
    p.children.length > 1 &&
    p.children[0].type === "linkReference" &&
    p.children[0].referenceType === "shortcut" &&
    p.children[0].identifier.match(timestampRegex)
  );
}

// allow no timestamp?
function isSpeaker(p) {
  return p.children.length > 1 && p.children[0].type === "strong";
}

// https://www.gatsbyjs.org/tutorial/remark-plugin-tutorial/
module.exports = ({ markdownAST }, pluginOptions) => {
  let firstSpeaker;
  visit(markdownAST, "paragraph", (node) => {
    if (node.data && (node.data.hName || node.data.hProperties)) return;

    let timestamp = isTimestamp(node) && isTimestamp(node)[0];
    if (timestamp) {
      node.children.shift(); // remove timestamp
      node.children.shift(); // remove space
    } else if (isSpeaker(node)) {
      timestamp = "00:00";
    } else {
      return;
    }
    let name = toString(node.children[0]);
    if (!firstSpeaker) {
      firstSpeaker = name;
    }

    node.children = [
      {
        type: "html",
        children: undefined,
        value: `<a href="#t=${timestamp}"><span class="timestamp">${timestamp} üïê</span></a>`,
      },
      {
        type: "paragraph",
        data: {
          hName: "span",
          hProperties: {
            className: ["text"],
          },
        },
        children: node.children,
      },
    ];

    node.type = "paragraph";
    node.data = {
      hProperties: {
        className: ["wrap"],
      },
    };
    if (name === firstSpeaker) {
      node.data.hProperties.className.push("right");
    }
  });

  return markdownAST;
};
