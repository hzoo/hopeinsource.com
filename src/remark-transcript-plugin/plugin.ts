import { visit } from "unist-util-visit";
import { toString } from "mdast-util-to-string";
import type { Plugin } from "unified";
import type { Paragraph, Text, Strong, Parent, PhrasingContent } from "mdast";

const timestampRegex = /^\[(\d{2}:\d{2})\]/;

interface PluginOptions {
  timestampClass?: string;
  wrapClass?: string;
  rightClass?: string;
  otherClass?: string;
  timestampEmoji?: string;
  textClass?: string;
}

function isTimestamp(node: Paragraph): boolean {
  if (node.children.length < 2) return false;
  const [first, second] = node.children;
  return (
    first.type === "text" &&
    second.type === "strong" &&
    timestampRegex.test((first as Text).value)
  );
}

function isSpeaker(node: Paragraph): boolean {
  return node.children.length > 0 && node.children[0].type === "strong";
}

interface HTMLStrong extends Parent {
  type: "strong";
  data: {
    hName: "strong";
  };
}

function createText(value: string): Text {
  return { type: "text", value };
}

function createSpan(
  className: string,
  children: (PhrasingContent | HTMLStrong | Text)[]
) {
  return {
    type: "span",
    data: {
      hName: "span",
      hProperties: { className: [className] },
    },
    children,
  };
}

function createStrong(children: Text[]): HTMLStrong {
  return {
    type: "strong",
    data: {
      hName: "strong",
    },
    children,
  };
}

export const remarkTranscriptPlugin: Plugin<[PluginOptions?]> = (
  options: PluginOptions = {}
) => {
  const {
    timestampClass = "timestamp",
    wrapClass = "wrap",
    rightClass = "right",
    otherClass = "other",
    timestampEmoji = "🕐",
    textClass = "text",
  } = options;

  return (tree) => {
    const speakerOrder: string[] = [];

    visit(tree, "paragraph", (node: Paragraph) => {
      let timestamp: string | null = null;
      let speaker: string | null = null;

      if (isTimestamp(node)) {
        const match = (node.children[0] as Text).value.match(timestampRegex);
        if (match) {
          timestamp = match[1];
          speaker = toString(node.children[1] as Strong);
        }
      } else if (isSpeaker(node)) {
        timestamp = "00:00";
        speaker = toString(node.children[0] as Strong);
      } else {
        return;
      }

      if (!timestamp || !speaker) return;

      if (!speakerOrder.includes(speaker)) {
        speakerOrder.push(speaker);
      }

      const timestampLink = {
        type: "link",
        url: `#t=${timestamp}`,
        children: [
          createSpan(timestampClass, [
            createText(`${timestamp} ${timestampEmoji}`),
          ]),
        ],
      };

      const content = isTimestamp(node)
        ? node.children.slice(2)
        : node.children.slice(1);

      const textSpan = createSpan(textClass, [
        createStrong([createText(`${speaker}`)]),
        ...content,
      ]);

      node.children = [
        timestampLink as PhrasingContent,
        textSpan as PhrasingContent,
      ];
      node.data = {
        hName: "p",
        hProperties: {
          className: [
            wrapClass,
            speakerOrder.indexOf(speaker) === 0 ? rightClass : otherClass,
          ],
        },
      };
    });
  };
};
