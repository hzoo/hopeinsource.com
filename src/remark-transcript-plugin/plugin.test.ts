import { test, expect } from "bun:test";
import { unified } from "unified";
import remarkParse from "remark-parse";
import remarkRehype from "remark-rehype";
import rehypeStringify from "rehype-stringify";

import { remarkTranscriptPlugin } from "./plugin";

const processMarkdown = async (markdown: string, options = {}) => {
  const result = await unified()
    .use(remarkParse)
    .use(remarkTranscriptPlugin, options)
    .use(remarkRehype)
    .use(rehypeStringify)
    .process(markdown);
  return result.toString();
};

test("remarkTranscriptPlugin transforms markdown correctly", async () => {
  const input = `
[00:28] **Speaker 1**: Hello, this is a test.

[01:15] **Speaker 2**: This is another test line.

**Speaker 1**: This is a line without a timestamp.
`;

  const expectedOutput = `<p class="wrap right"><a href="#t=00:28"><span class="timestamp">00:28 🕐</span></a><span class="text"><strong>Speaker 1</strong>: Hello, this is a test.</span></p>
<p class="wrap other"><a href="#t=01:15"><span class="timestamp">01:15 🕐</span></a><span class="text"><strong>Speaker 2</strong>: This is another test line.</span></p>
<p class="wrap right"><a href="#t=00:00"><span class="timestamp">00:00 🕐</span></a><span class="text"><strong>Speaker 1</strong>: This is a line without a timestamp.</span></p>`;

  const result = await processMarkdown(input);
  expect(result.trim()).toBe(expectedOutput.trim());
});

test("remarkTranscriptPlugin handles empty input", async () => {
  const input = "";
  const result = await processMarkdown(input);
  expect(result).toBe("");
});

test("remarkTranscriptPlugin ignores non-transcript paragraphs", async () => {
  const input = "This is a regular paragraph without timestamps or speakers.";
  const result = await processMarkdown(input);
  expect(result).toBe("<p>" + input + "</p>");
});

test("remarkTranscriptPlugin handles custom options", async () => {
  const input = "[00:28] **Speaker**: Test with custom options.";

  const result = await processMarkdown(input, {
    timestampClass: "custom-timestamp",
    timestampEmoji: "⏰",
  });

  expect(result).toContain('class="custom-timestamp"');
  expect(result).toContain("⏰");
});
