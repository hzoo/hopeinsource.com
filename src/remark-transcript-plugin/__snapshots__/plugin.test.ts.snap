// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't affect normal paragraphs or use of link references 1`] = `
{
  "children": [
    {
      "type": "text",
      "value": "hi [asdf]",
    },
  ],
  "type": "paragraph",
}
`;

exports[`adds id to a markdown header 1`] = `
{
  "children": [
    {
      "type": "text",
      "value": "[00:00] ",
    },
    {
      "children": [
        {
          "type": "text",
          "value": "Henry",
        },
      ],
      "type": "strong",
    },
    {
      "type": "text",
      "value": ": Hi.",
    },
  ],
  "type": "paragraph",
}
`;
